### ISSUE TITLE
Different results for forward pass of two equal tensors through Conv2d

### ISSUE THREAD

**Issue by dozed** [pytorch/pytorch#110387]
### 🐛 Describe the bug

Forwarding a tensor `img` through a simple PyTorch Conv2d model produces a different result than forwarding `img + torch.zeros_like(img)`.

Here is a minimal example: https://github.com/dozed/pytorch-issue-1/blob/main/test_issue.py

```python
# prepare input
img = torchvision.io.image.read_image('sky1024px.jpg')
img = FT.convert_image_dtype(img, torch.float32)
img = img.unsqueeze(dim=0)

# prepare input + zero
zeros = torch.zeros_like(img)
img_updated = img + zeros

# input tensors are identical
assert torch.allclose(img, img_updated)
assert torch.equal(img, img_updated)

# prepare model
conv1 = nn.Conv2d(in_channels=3, out_channels=129, kernel_size=3, padding=1)
conv2 = nn.Conv2d(in_channels=129, out_channels=4, kernel_size=1)

# forward 1
x = conv1(img)
result1 = conv2(x)

# forward 2
y = conv1(img_updated)
result2 = conv2(y)

# tensors after conv1 are equal
assert torch.allclose(x, y)
assert torch.equal(x, y)

# ISSUE: the results are not equal but should be, since only zeros are added
print(torch.linalg.norm(result1 - result2))
assert torch.allclose(result1, result2)
assert torch.equal(result1, result2)
```

### Versions

```
PyTorch version: 2.0.1
Is debug build: False
CUDA used to build PyTorch: None
ROCM used to build PyTorch: N/A

OS: Ubuntu 23.04 (x86_64)
GCC version: (Ubuntu 12.3.0-1ubuntu1~23.04) 12.3.0
Clang version: Could not collect
CMake version: version 3.25.1
Libc version: glibc-2.37

Python version: 3.11.5 (main, Sep 11 2023, 13:54:46) [GCC 11.2.0] (64-bit runtime)
Python platform: Linux-6.2.0-32-generic-x86_64-with-glibc2.37
Is CUDA available: False
CUDA runtime version: No CUDA
CUDA_MODULE_LOADING set to: N/A
GPU models and configuration: No CUDA
Nvidia driver version: No CUDA
cuDNN version: No CUDA
HIP runtime version: N/A
MIOpen runtime version: N/A
Is XNNPACK available: True

CPU:
Architecture:                       x86_64
CPU op-mode(s):                     32-bit, 64-bit
Address sizes:                      46 bits physical, 48 bits virtual
Byte Order:                         Little Endian
CPU(s):                             12
On-line CPU(s) list:                0-11
Vendor ID:                          GenuineIntel
Model name:                         12th Gen Intel(R) Core(TM) i7-1265U
CPU family:                         6
Model:                              154
Thread(s) per core:                 2
Core(s) per socket:                 10
Socket(s):                          1
Stepping:                           4
CPU(s) scaling MHz:                 37%
CPU max MHz:                        4800,0000
CPU min MHz:                        400,0000
BogoMIPS:                           5376,00
Flags:                              fpu vme de pse tsc msr pae mce cx8 apic sep mtrr pge mca cmov pat pse36 clflush dts acpi mmx fxsr sse sse2 ss ht tm pbe syscall nx pdpe1gb rdtscp lm constant_tsc art arch_perfmon pebs bts rep_good nopl xtopology nonstop_tsc cpuid aperfmperf tsc_known_freq pni pclmulqdq dtes64 monitor ds_cpl vmx smx est tm2 ssse3 sdbg fma cx16 xtpr pdcm sse4_1 sse4_2 x2apic movbe popcnt tsc_deadline_timer aes xsave avx f16c rdrand lahf_lm abm 3dnowprefetch cpuid_fault epb ssbd ibrs ibpb stibp ibrs_enhanced tpr_shadow vnmi flexpriority ept vpid ept_ad fsgsbase tsc_adjust bmi1 avx2 smep bmi2 erms invpcid rdseed adx smap clflushopt clwb intel_pt sha_ni xsaveopt xsavec xgetbv1 xsaves split_lock_detect avx_vnni dtherm ida arat pln pts hwp hwp_notify hwp_act_window hwp_epp hwp_pkg_req hfi umip pku ospke waitpkg gfni vaes vpclmulqdq tme rdpid movdiri movdir64b fsrm md_clear serialize pconfig arch_lbr ibt flush_l1d arch_capabilities
Virtualization:                     VT-x
L1d cache:                          352 KiB (10 instances)
L1i cache:                          576 KiB (10 instances)
L2 cache:                           6,5 MiB (4 instances)
L3 cache:                           12 MiB (1 instance)
NUMA node(s):                       1
NUMA node0 CPU(s):                  0-11
Vulnerability Gather data sampling: Not affected
Vulnerability Itlb multihit:        Not affected
Vulnerability L1tf:                 Not affected
Vulnerability Mds:                  Not affected
Vulnerability Meltdown:             Not affected
Vulnerability Mmio stale data:      Not affected
Vulnerability Retbleed:             Not affected
Vulnerability Spec store bypass:    Mitigation; Speculative Store Bypass disabled via prctl
Vulnerability Spectre v1:           Mitigation; usercopy/swapgs barriers and __user pointer sanitization
Vulnerability Spectre v2:           Mitigation; Enhanced IBRS, IBPB conditional, RSB filling, PBRSB-eIBRS SW sequence
Vulnerability Srbds:                Not affected
Vulnerability Tsx async abort:      Not affected

Versions of relevant libraries:
[pip3] mypy==1.4.1
[pip3] mypy-extensions==0.4.3
[pip3] numpy==1.24.3
[pip3] pytorch-caffe-models==0.1
[pip3] torch==2.0.1
[pip3] torchdata==0.6.1
[pip3] torchtext==0.15.2
[pip3] torchvision==0.15.2
[pip3] torchviz==0.0.2
[conda] blas                      1.0                         mkl  
[conda] ffmpeg                    4.3                  hf484d3e_0    pytorch
[conda] mkl                       2023.1.0         h213fc3f_46343  
[conda] mkl-service               2.4.0           py311h5eee18b_1  
[conda] mkl_fft                   1.3.6           py311ha02d727_1  
[conda] mkl_random                1.2.2           py311ha02d727_1  
[conda] numpy                     1.24.3          py311h08b1b3b_1  
[conda] numpy-base                1.24.3          py311hf175353_1  
[conda] pytorch                   2.0.1              py3.11_cpu_0    pytorch
[conda] pytorch-caffe-models      0.1                      pypi_0    pypi
[conda] pytorch-mutex             1.0                         cpu    pytorch
[conda] torchdata                 0.6.1                    pypi_0    pypi
[conda] torchtext                 0.15.2                   pypi_0    pypi
[conda] torchvision               0.15.2                py311_cpu    pytorch
[conda] torchviz                  0.0.2                    pypi_0    pypi
```

**Comment by dozed** [comment]
Interestingly when changing `out_channels=129` to `out_channels=128` (and adjusting `in_channels` of the second Conv2d layer accordingly) the results are the same.

Also when forwarding a random image, e.g. `torch.randn((1, 3, 32, 32))` the results are the same as well.

**Comment by colesbury** [comment]
I don't think this is a bug. `result1` and `result2` are close but not identical, right? The convolution implementation may use a different kernel depending on if the input is contiguous or not. `img + zeros` makes the input contiguous.

**Comment by dozed** [comment]
Yep, they are close, `torch.allclose` passes and `torch.equal` not. In practice it would be good enough, I am just trying to understand what is happening when adding a `0`.

`img + zeros` btw does not make the input contiguous:

```python
def test_add_zero_different_result_2():
    set_seed(42)
    torch.set_printoptions(precision=9)

    # prepare non-contiguous input
    img = torch.tensor(
        [
            [[1.0, 1.0, 1.0]],
            [[1.0, 1.0, 1.0]],
            [[1.0, 1.0, 1.0]],
        ],
        dtype=torch.float32
    ).permute([2, 1, 0])

    img = img.unsqueeze(dim=0)

    zeros = torch.zeros_like(img)
    img_updated = img + zeros
    assert not img.is_contiguous()
    assert not img_updated.is_contiguous()

    conv1 = nn.Conv2d(in_channels=3, out_channels=2, kernel_size=3, padding=1)
    conv1.requires_grad_(False)

    x = conv1(img)
    y = conv1(img_updated)
    assert x.is_contiguous()
    assert not y.is_contiguous()

    print(torch.linalg.norm(x - y))  # 6.0069e-08
    assert torch.allclose(x, y)
    assert torch.equal(x, y)  # assertion fails
```

Interestingly `x` is contiguous and `y` is not contiguous, even if `img` and `img_updated` both are not contiguous.

**Comment by colesbury** [comment]
The stride generation is not the problem. Strides for size 1 dimensions are not meaningful. The detection of `ChannelsLast` should not consider strides for dimensions of size 1.

But in general, you should still not assume that `f(input + zero)` and `f(input)` produce the bitwise exact result because adding zero to a tensor can change the layout of the result.

**Comment by dozed** [comment]
@colesbury The memory layout detection takes into account the stride of the first dimension, e.g. <https://github.com/pytorch/pytorch/blob/main/c10/core/MemoryFormat.h#L142>:
- For sizes `[1, 3, 1, 2]` and (incorrect) strides `[3, 1, 6, 3]` this will return `false`
- For sizes `[1, 3, 1, 2]` and (correct) strides `[6, 1, 6, 3]` this will return `true`

Here is a minimal working example: https://github.com/dozed/pytorch-issue-3/blob/main/main.cpp

**Comment by colesbury** [comment]
Maybe @ezyang can weigh in.

**Comment by ezyang** [comment]
I agree with the classification of incorrect/correct strides. My inclination would be to beef up the unsqueeze stride selection heuristic to prefer the MAX stride, rather than the leftmost one. @dozed do you think you'd be able to send a patch, so we can give it a whirl in CI and see what happens?

**Comment by dozed** [comment]
Sure I can give it a try. Is there an existing maximum function which operates over an `at::IntArrayRef`?

**Comment by ezyang** [comment]
Probably not

**Comment by dozed** [comment]
@ezyang By experimenting a bit I found that $stride_{new} = \max_i stride[i]$ does not always hold, e.g.:

Tensor $x$
- $size = [2, 2, 3]$ and
- $stride = [6, 3, 1]$

Tensor $y = x.permute(2, 1, 0)$
- $size = [3, 2, 2]$ and
- $stride = [1, 3, 6]$

Tensor $z = y.unsqueeze(dim=0)$ gives:
- $size = [1, 3, 2, 2]$ and
- $stride = [3, 1, 3, 6]$

Remarks:
- The currently computed $stride[0] = 3$ for tensor $z$ is incorrect here
- $max_i stride[i] = 6$ is also incorrect
- Correct would be $max_i stride[i] \cdot size[i] = 12$

So:
- For $dim = 0$ the rule $stride_{new} = \max_i stride[i] \cdot size[i]$ could be used.
- For $dim >= t.dim()$, $stride_{new} = 1$ is fine (as it is currently implemented as well).
- For $0 < dim < t.dim()$ things could get tricky. I wanted to experiment a bit more.

A simple first solution would be to extend the current $unsqueeze$ heuristic for the case $dim=0$. What do you think?

**Comment by ezyang** [comment]
That also seems fine.

**Comment by dozed** [comment]
I have now two alternative patch options:
- Patch 1: modifies unsqueeze stride heuristic
- Patch 2: modifies memory layout check (`is_channels_last_strides_2d_s4`)

# Patch 1

As discussed for $dim = 0: stride_{new} = \max_i stride[i] \cdot size[i]$:

```diff
diff --git a/aten/src/ATen/native/TensorShape.cpp b/aten/src/ATen/native/TensorShape.cpp
index 0a018fbc8db..683b61e9e45 100644
--- a/aten/src/ATen/native/TensorShape.cpp
+++ b/aten/src/ATen/native/TensorShape.cpp
@@ -3188,10 +3188,26 @@ struct InferUnsqueezeGeometryResult {
       : sizes(tensor_sizes.begin(), tensor_sizes.end())
       , strides(tensor_strides.begin(), tensor_strides.end()) {}
 };
+
+int64_t get_first_dim_stride_for_unsqueeze(const InferUnsqueezeGeometryResult& g) {
+  assert(!g.sizes.empty());
+  int64_t max_stride = 0;
+
+  for (int64_t i = 0; i < g.sizes.size(); i++) {
+    int64_t candidate_stride = g.sizes.at(i) * g.strides.at(i);
+
+    if (candidate_stride > max_stride) {
+      max_stride = candidate_stride;
+    }
+  }
+
+  return max_stride;
+}
+
 InferUnsqueezeGeometryResult
 inferUnsqueezeGeometry(const Tensor& tensor, int64_t dim) {
   InferUnsqueezeGeometryResult result(tensor.sizes(), tensor.strides());
-  int64_t new_stride = dim >= tensor.dim() ? 1 : result.sizes[dim] * result.strides[dim];
+  int64_t new_stride = dim >= tensor.dim() ? 1 : (dim == 0 ? get_first_dim_stride_for_unsqueeze(result) : result.sizes[dim] * result.strides[dim]);
   result.sizes.insert(result.sizes.begin() + dim, 1);
   result.strides.insert(result.strides.begin() + dim, new_stride);
```

# Patch 2

Ignores strides for $size=1$ dimensions:

```diff
diff --git a/c10/core/MemoryFormat.h b/c10/core/MemoryFormat.h
index f4e3e930790..61cc721c56a 100644
--- a/c10/core/MemoryFormat.h
+++ b/c10/core/MemoryFormat.h
@@ -138,7 +138,7 @@ inline bool is_channels_last_strides_2d_s4(
     if (sizes[d] == 0) {
       return false;
     }
-    if (strides[d] < min) {
+    if (strides[d] < min && sizes[d] != 1) {
       return false;
     }
     // Fallback to NCHW as default layout for ambiguous cases
```

Any preference @ezyang?

Maybe @jjsjann123 can give some feedback on this issue as well (he authored `is_channels_last_strides`)?

**Comment by jjsjann123** [comment]
I think going with `Patch 1` sounds like makes more sense to me.

I agree that stride for dimension with `size == 1` isn't semantically meaningful, but implicit stride order does rely on it to propagate the intended stride order.

Especially with the change proposed here. We'll just never go through channels last when we have a dimension with stride == 0. We'll be missing out on performance in that case.

cc'ing @ptrblck

**Comment by dozed** [comment]
Ok, thanks :) I will then prepare a pull request for option 1.

Does https://github.com/pytorch/pytorch/blob/main/torch/_meta_registrations.py#L380-L386 needs to be modified as well?

For completeness: NumPy has the same behaviour as PyTorch currently. I reported the issue in https://github.com/numpy/numpy/issues/25225 but they wont fix it.

**Comment by ptrblck** [comment]
> Especially with the change proposed here. We'll just never go through channels last when we have a dimension with stride == 0. We'll be missing out on performance in that case.

So are trading performance for what exactly? To disallow different code paths introducing expected numerical mismatches caused by the limited floating point precision?

> For completeness: NumPy has the same behaviour as PyTorch currently. I reported the issue in numpy/numpy#25225 but they wont fix it.

So why do we want to diverge and introduce performance regressions? 

CC @malfet

**Comment by dozed** [comment]
@ptrblck The issue diverged a bit from avoiding numerical mismatches to having correct strides and helping the memory layout inference procedure.

I am not sure how important it is to let the memory layout inference procedure correctly determine `ChannelsLast` memory layout. So I let you guys decide :)

**Comment by ptrblck** [comment]
> I am not sure how important it is to let the memory layout inference procedure correctly determine ChannelsLast memory layout. 

Channels-last is especially important for mixed-precision types and according to @jjsjann123 the proposed change would disable channels-last code paths potentially introducing performance regressions.

>  to having correct strides and helping the memory layout inference procedure.

I'm sure I'm missing important details, but don't fully understand how the two previously mentioned drawbacks compare to your proposal of having "correct strides" and what it exactly means given we are also diverging from numpy.

**Comment by ezyang** [comment]
So, to be clear, the reason why I'm willing to go along with the proposed changes here, is not because I care about the change in convolution precision, but because it seems reasonable to think about what we want the intended contiguity change for unsqueeze should be when you have a channels-last tensor.

Specifically, suppose I have a tensor in CHW format. I wish to unsqueeze it to give N=1 batch dimension. Today, unsqueeze(0) will give me a tensor with strides (HW, W, 1) which is fine and intuitive.

```
>>> torch.empty_permuted([2, 3, 5], [0, 1, 2]).unsqueeze(0).stride()
(30, 15, 5, 1)
```

But now suppose I had a HWC tensor. Now unsqueeze(0) doesn't give me an NHWC tensor:

```
>>> torch.empty_permuted([2, 3, 5], [1, 2, 0]).stride()
(1, 10, 2)
>>> torch.empty_permuted([2, 3, 5], [1, 2, 0]).unsqueeze(0).stride()
(2, 1, 10, 2)
```

Instead of 2, I would have preferred the stride to be 30. So @dozed is trying to devise a modified heuristic for unsqueeze that will make this happen. Option 1 seems pretty reasonable to me.

It is true that diverging from Numpy stride behavior on unsqueeze is a bit insalubrious. But Numpy doesn't care what this stride is, right? PyTorch does, because we have memory layout propagation rules. So I think I would still like to see this PR. It may not work, but it will be informative why it doesn't work.
